src/program_generator/pg_marginal.py

```diff
@@ -89,18 +91,82 @@ def _do_generate(self, opcode, max_op_count, shuffle_counts, step_op_count):
     if shuffle_counts:
       random.shuffle(op_counts)
       
-    programs = [self._generate_single_program(operation, op_count) for operation in operations for op_count in op_counts]
+    programs = [self._generate_single_program(operation, op_count, max_op_count) for operation in operations for op_count in op_counts]
 
     return programs
 
-  def _generate_single_program(self, operation, op_count):
+  def _generate_single_program(self, operation, op_count, max_op_count):
+    assert op_count <= constants.MAX_INSTRUCTIONS
     # for compatibility with the generate_single_marginal function
     if operation['Mnemonic'] == 'CREATE':
-        return Program(generate_create_program(operation, op_count), operation['Mnemonic'], op_count)
-
+      return Program(_generate_create_program(operation, op_count, max_op_count), operation['Mnemonic'], op_count)
+    if operation['Mnemonic'] in ['EXTCODEHASH', 'EXTCODESIZE']:
+      return Program(_generate_extcode_program(operation, op_count, max_op_count), operation['Mnemonic'], op_count)
+    if operation['Mnemonic'] == 'EXTCODECOPY':
+      return Program(_generate_extcodecopy_program(operation, op_count, max_op_count), operation['Mnemonic'], op_count)
+    if operation['Mnemonic'] in ['CALL', 'STATICCALL', 'DELEGATECALL']:
+      return Program(_generate_call_program(operation, op_count, max_op_count), operation['Mnemonic'], op_count)
+  
     single_op_pushes = ["6003"] * arity(operation)
 
     return Program(generate_single_marginal(single_op_pushes, operation, op_count), operation['Mnemonic'], op_count)
+
+def _generate_create_program(create_operation, op_count, max_op_count):
```

Comment:
Moved here, because it shouldn't be in `common.py`. It's not used outside this file.